### ISSUE TITLE
Latest versions of Concourse are uploded to Go index servers with versions 1.6.1

### ISSUE THREAD

**Issue by yurivict** [concourse/concourse#7109]
wget "https://index.golang.org/index?limit=1&since=2021-05-28T01:21:42.994935Z" returns version  ```v1.6.1-0.20210527214455```.

**Comment by taylorsilva** [comment]
Is this similar to https://github.com/concourse/concourse/issues/3952?

**Comment by yurivict** [comment]
Perhaps.

* The URL above returns a wrong Concourse version: ```v1.6.1-0.20210527214455```.
* This URL https://pkg.go.dev/github.com/concourse/concourse shows: ```v4,2,3```
* GitHub repository shows ```7.3.1```

When I try to update the FreeBSD port it tries to pull the correct version ```7.4.1``` from Go's servers and it doesn't exist there.

Not sure how this works. Did you ever register the project on https://pkg.go.dev/github.com/concourse/concourse? Maybe you need to correct it there.

**Comment by yurivict** [comment]
https://github.com/golang/go/issues/46494

**Comment by taylorsilva** [comment]
I don't think we ever did register it on there. The readme on that page looks **really** old. If you read #3952 you'll see that the main repo isn't designed as a go library/package at all. It's not meant to be imported by other go applications and we don't plan to support that workflow in general. 

Sounds like your usecase is different though and reminds me of #6712 (hey look, it's you again!).

**Comment by yurivict** [comment]
This has nothing to do with concourse being imported as library,

The FreeBSD ports framework uses  pkg.go.dev -based URLs to retrieve package description with required module information, etc. The standard URL to get this file doesn't work for concourse due to version mismatch.

**Comment by yurivict** [comment]
Go responded:

```
+incompatible is for repos that have not adopted modules (no go.mod) The last version this was true for github.com/concourse/concourse is v4.2.3
concourse is using modules (go.mod) but not versioning in accordance with policy, so the version numbers are pseudoversions

pkg.go.dev will only list valid versions, though every pseudoversion is also available if you consttuct the url

closing as this is working as intended
```

**Comment by taylorsilva** [comment]
> The FreeBSD ports framework uses pkg.go.dev -based URLs to retrieve package description with required module information, etc. The standard URL to get this file doesn't work for concourse due to version mismatch.

I don't know what's involved in this process. My naive suggestion is to track the tags on our repo here. I'm guessing that doesn't give you all the info you need though :(

**Comment by mmwtsn** [comment]
> Is this similar to #3952?

Yes. That issue is about the Concourse module being invalid. This issue is a symptom of the invalid module.

The module is invalid because the module path in `go.mod` does not contain a major version but the git tag does:

```sh
$ GOPROXY=proxy.golang.org go list -m github.com/concourse/concourse@v7.8.2                                                                                                                                                                             
go: github.com/concourse/concourse@v7.8.2: reading https://proxy.golang.org/github.com/concourse/concourse/@v/v7.8.2.info: 404 Not Found                                                                                                                                               
        server response: not found: github.com/concourse/concourse@v7.8.2: invalid version: module contains a go.mod file, so module path must match major version ("github.com/concourse/concourse/v7")
```

Go modules consider new semver major versions to be backwards-incompatible with previous versions and therefore expects each new major version to have a separate module path. The latest git tag is `v7.8.2` but the module definition in `go.mod` has not been updated since v0 or v1. There are [two general strategies](https://go.dev/blog/v2-go-modules#major-version-strategies) to resolve that:

1. Publish all versions off a main branch and move new versions into `vX` directories
2. Publish each version separately off a release branch and rename the module

It's straightforward to rename the module paths, but that is potentially a breaking change for anyone who is importing this module right now, which I think is only possible using a branch or commit hash directly, as the release is invalid:

```sh
$ go get github.com/concourse/concourse@v7.8.2
go: github.com/concourse/concourse@v7.8.2: invalid version: module contains a go.mod file, so module path must match major version ("github.com/concourse/concourse/v7")

$ go get github.com/concourse/concourse@31aae53
go: downloading github.com/concourse/concourse v1.6.1-0.20220714181844-31aae53c77f8

$ go get github.com/concourse/concourse@master
go: upgraded github.com/concourse/concourse v1.6.1-0.20220714181844-31aae53c77f8 => v1.6.1-0.20220907210330-17dbb45a7667
```

Those situations are arguably relying on broken behavior though. The two successful `go get` runs above downloaded a pre-release version of `v1.6.1` with the format `v1.6.1-0.<timestamp>-<commit>` which actually has the contents of `master` and the `v7.8.2` tag from what I can tell in my local module cache. That patch release should not be v7.

How does the team feel about releasing this as `v8.0.0` and a `/v8` module path versus a `v7.x.x` patch release? I have branches ready with both changes locally and I can open a pull request for either. I assume anyone using the modules is not going to need the previous major versions updated as well, but if we go the v7 patch approach, we could do that.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
